### ISSUE TITLE
[stdlib] Use `Intable` in `getitem` and `setitem` consistently

### ISSUE THREAD

**Issue by JoeLoser** [modular/modular#2337]
Many `__getitem__` and `__setitem__` functions still take an `Int` as an argument.  This is OK, but forces call sites to explicitly construct an `Int` via `int(my_value)` for example to call `__getitem__` or `__setitem__`.  Instead, we can just make `__getitem__` and `__setitem__` take an `Intable` thing that always does the `int(value)` internally rather than pushing it onto all the callers.

One thing to note is for floating point indices, like `x[0.1]`, this would erroneously call `x[int(0.1)]` i.e. `x[0]`.  This is surprising and instead should be a hard error.  But we can put that on hold and stage this in phases:

1. Change all the `__getitem__` and `__setitem__` functions to work on `Intable` and ignore the floating point indexing issue mentioned above.
2. Introduce an `Integral` trait or something to ensure that callers can't pass a floating point value to these functions.  Use that instead of `Intable` everywhere.
3. Audit other functions using `Intable` in case they also have floating point indexing issues like mentioned above.

FYI @itramble

**Comment by gabrieldemarmiesse** [comment]
The python community already solved this issue of floating points :) See https://docs.python.org/3/reference/datamodel.html#object.__index__

This is what is called internally when doing `my_list[some_object]`, this is also why `my_list[4.2]` won't work in python.

**Comment by gabrieldemarmiesse** [comment]
@bgreni I don't think you need to ask permission to work on an issue. It's quite uncommon for two contributors to work on the same thing. If you don't see a PR open, it's likely that no one is working on it :)

**Comment by lattner** [comment]
@JoeLoser I believe we turned around and now want everything to use Int consistently instead of Intable.  Do you agree?  If so, we should tidy up this issue to avoid confusion

**Comment by soraros** [comment]
@JoeLoser I think this is fixed given we use `Indexer` now?

**Comment by soraros** [comment]
@JoeLoser Gentle ping.

**Comment by JoeLoser** [comment]
Agreed, closing!
